lib/ansible/modules/cloud/vmware/vmware_host.py

```diff
@@ -107,99 +114,107 @@
+            self.module.fail_json(msg=to_native(runtime_fault.msg))
         except vmodl.MethodFault as method_fault:
-            self.module.fail_json(msg=method_fault.msg)
+            self.module.fail_json(msg=to_native(method_fault.msg))
         except Exception as e:
-            self.module.fail_json(msg=str(e))
+            self.module.fail_json(msg=to_native(e))
 
     def add_host_to_vcenter(self):
-
-        if self.esxi_username is None or self.esxi_password is None:
-            self.module.fail_json(msg='esxi_username and esxi_password are required to add a host')
-
         host_connect_spec = vim.host.ConnectSpec()
         host_connect_spec.hostName = self.esxi_hostname
         host_connect_spec.userName = self.esxi_username
         host_connect_spec.password = self.esxi_password
         host_connect_spec.force = True
-        host_connect_spec.sslThumbprint = ""
+        host_connect_spec.sslThumbprint = self.esxi_ssl_thumbprint
         as_connected = True
         esxi_license = None
         resource_pool = None
 
-        try:
-            task = self.cluster.AddHost_Task(host_connect_spec, as_connected, resource_pool, esxi_license)
-            success, result = wait_for_task(task)
-            return success, result
-        except TaskError as add_task_error:
-            # This is almost certain to fail the first time.
-            # In order to get the sslThumbprint we first connect
-            # get the vim.fault.SSLVerifyFault then grab the sslThumbprint
-            # from that object.
-            #
-            # args is a tuple, selecting the first tuple
-            ssl_verify_fault = add_task_error.args[0]
-            host_connect_spec.sslThumbprint = ssl_verify_fault.thumbprint
+        if self.esxi_ssl_thumbprint == '':
+            # User has not specified SSL Thumbprint for ESXi host,
+            # try to grab it using SSLVerifyFault exception
+            try:
+                task = self.cluster.AddHost_Task(host_connect_spec, as_connected, resource_pool, esxi_license)
+                success, result = wait_for_task(task)
+                return success, result
+            except TaskError as e:
+                self.module.fail_json(msg="Failed to add host %s to vCenter due to %s" % (self.esxi_hostname, e))
+            except vim.fault.SSLVerifyFault as add_task_error:
+                # This is almost certain to fail the first time.
+                # In order to get the sslThumbprint we first connect
+                # get the vim.fault.SSLVerifyFault then grab the sslThumbprint
+                # from that object.
+                #
+                # args is a tuple, selecting the first tuple
+                ssl_verify_fault = add_task_error.args[0]
+                host_connect_spec.sslThumbprint = ssl_verify_fault.thumbprint
 
         task = self.cluster.AddHost_Task(host_connect_spec, as_connected, resource_pool, esxi_license)
         success, result = wait_for_task(task)
```

Comment:
I would have write:
```
-        try:
-            task = self.cluster.AddHost_Task(host_connect_spec, as_connected, resource_pool, esxi_license)
-            success, result = wait_for_task(task)
-            return success, result
-        except TaskError as add_task_error:
-            # This is almost certain to fail the first time.
-            # In order to get the sslThumbprint we first connect
-            # get the vim.fault.SSLVerifyFault then grab the sslThumbprint
-            # from that object.
-            #
-            # args is a tuple, selecting the first tuple
-            ssl_verify_fault = add_task_error.args[0]
-            host_connect_spec.sslThumbprint = ssl_verify_fault.thumbprint
-
-        task = self.cluster.AddHost_Task(host_connect_spec, as_connected, resource_pool, esxi_license)
-        success, result = wait_for_task(task)
-        return success, result
+        for _ in range(0, 2):
+            try:
+                task = self.cluster.AddHost_Task(host_connect_spec, as_connected, resource_pool, esxi_license)
+                success, result = wait_for_task(task)
+                return success, result
+
+            except TaskError as task_error_exception:
+
+                task_error = task_error_exception.args[0]
+                if self.esxi_ssl_thumbprint == '' and isinstance(task_error, vim.fault.SSLVerifyFault):
+                    # User has not specified SSL Thumbprint for ESXi host,
+                    # try to grab it using SSLVerifyFault exception
+                    host_connect_spec.sslThumbprint = task_error.thumbprint
+                else:
+                    self.module.fail_json(msg="Failed to add host %s to vCenter: %s" % (self.esxi_hostname, task_error.msg))
+
+        self.module.fail_json(msg="Failed to add host %s to vCenter: %s" % (self.esxi_hostname, task_error.msg))
```

It seems vim.fault.SSLVerifyFault and other vim.fault… are always packed in a TaskError.

Reply:
@Akasurde what do you think ?